File: moodstocks-api/msapi.rb
# NOTE: get rufus-verbs at https://github.com/catwell/rufus-verbs

%w{rubygems rufus-verbs rest-client uri json}.each{|x| require x}

# Settings
@@key = "YourApiKey"
@@secret = "YourApiSecret"
image_filename = "sample.jpg"
image_url = "http://api.moodstocks.com/static/sample-book.jpg"
id = "test1234"

class MSApi; include Rufus::Verbs

  attr_reader :ep

  def initialize
    @ep = EndPoint.new({
      host: "api.moodstocks.com",
      port: 80,
      resource: "v2",
      digest_authentication: [@@key,@@secret],
    })
  end

  def disp(r)
    if r.code.to_i == 200
      puts "[OK]\t#{JSON.parse(r.body)}"
    else
      puts "[ERROR]\t#{r.code} #{r.body}"
    end
  end

  def do(verb,resource,pp={})
    disp ep.send(verb,pp.merge(id: resource))
  end

end

MS = MSApi.new

# Authenticating with your API key (Echo service)
MS.do(:get,"echo",{params:{query:{foo:"bar",bacon:"chunky"}}})

# Adding a reference image
body,headers = File.open(image_filename,'rb') do |f|
  mp = RestClient::Payload::Multipart.new(image_file:f)
  [mp.read,mp.headers]
end
imgdata = {data:body,headers:headers}
MS.do(:put,"ref/#{id}",imgdata)

# Making an image available offline
MS.do(:post,"ref/#{id}/offline",{data:""})

# Using online search
MS.do(:post,"search",imgdata)

# Updating a reference & using a hosted image
MS.do(:put,"ref/#{id}",{data:"",query:{image_url:image_url}})

# Removing reference images
MS.do(:delete,"ref/#{id}")

File: moodstocks-api/msapi.php
<?php

// Settings
$key = "YourApiKey";
$secret = "YourApiSecret";
$image_filename = "sample.jpg";
$image_url = "http://api.moodstocks.com/static/sample-book.jpg";
$id = "test1234";

// CURL
$curl_opts = array(
  CURLOPT_RETURNTRANSFER=>true,
  CURLOPT_HTTPAUTH=>CURLAUTH_DIGEST,
  CURLOPT_USERPWD=>$key.":".$secret
);
$api_ep = "http://api.moodstocks.com/v2";

function disp($opts){
  $ch = curl_init();
  curl_setopt_array($ch, $opts);
  $raw_resp = curl_exec($ch);
  $array_resp = json_decode($raw_resp);
  print_r($array_resp);
  curl_close($ch);
}

// Authenticating with your API key (Echo service)
$params = array("foo"=>"bar","bacon"=>"chunky");
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/echo?".http_build_query($params);
disp($opts);

// Adding a reference image
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/ref/".$id;
$opts[CURLOPT_POSTFIELDS] = array("image_file"=>"@".$image_filename);
$opts[CURLOPT_CUSTOMREQUEST] = "PUT";
disp($opts);

// Making an image available offline
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/ref/".$id."/offline";
$opts[CURLOPT_CUSTOMREQUEST] = "POST";
disp($opts);

// Using online search
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/search";
$opts[CURLOPT_POSTFIELDS] = array("image_file"=>"@".$image_filename);
disp($opts);

// Updating a reference & using a hosted image
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/ref/".$id;
$opts[CURLOPT_POSTFIELDS] = array("image_url"=>$image_url);
$opts[CURLOPT_CUSTOMREQUEST] = "PUT";
disp($opts);

// Removing reference images
$opts = $curl_opts;
$opts[CURLOPT_URL] = $api_ep."/ref/".$id;
$opts[CURLOPT_CUSTOMREQUEST] = "DELETE";
disp($opts);

?>
